[Personal WIP] Story of Layout feature

albertusdev opened the pull request:
A PR to show my progress on working on this feature.

Dependency:
- Flutter framework change:
https://github.com/albertusangga/flutter/pull/2

Commit:
Show constraints next to Widget in Summary Tree

albertusangga commented:
Very early prototype on showing Constraints next to the Widget:

![image](https://user-images.githubusercontent.com/25892469/67442309-18795000-f5b5-11e9-9065-460c15253de5.png)

Commit:
Highlight children's constraints that got unconstrained, highlight parent widget that did the unconstraining, and add animation

albertusangga commented:
Check-in #2:
- Rename Show Constraints button to Layout button (change term of showConstraints to debugLayoutModeEnabled)
- Change color from red to yellow (Dangerous to Warning, since we are unsure whether it's really an error)
- Highlight children that got unconstrained by its parents (highlight the contraints text)
- Highlight the parent widget name that unconstrained its child
- Show fade-in/fade-out animation when Debug Layout mode 

![image](https://user-images.githubusercontent.com/25892469/67453643-2393a680-f5dc-11e9-9248-0fcffb6e0afd.png)

albertusangga commented:
Check-in 3:
- Made very early prototype on explaining Flex layouting (the feature for now will be temporarily called 'Story of Your Flex Widget')
- Visualizing Flex and able to detect the Flex type (Row/Column)
- Displaying correct amount of children and all of the children's widget name
- All children is still equally weighted for this check in

![image](https://user-images.githubusercontent.com/25892469/67453767-9270ff80-f5dc-11e9-8179-50b1e81016c8.png)
![image](https://user-images.githubusercontent.com/25892469/67453742-8127f300-f5dc-11e9-8b62-265ddfc855a9.png)

Commit:
Very early prototype of 'Story of Your Flex Widget' dialog

jacob314 commented on packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart:
```diff
@@ -4,6 +4,7 @@
 
 import 'dart:math';
 
+import 'package:devtools_app/src/ui/fake_flutter/_real_flutter.dart';
```
auto imports are doing the wrong thing. just go ahead and import 
'package:flutter/material.dart'

jacob314 commented on packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart:
```diff
@@ -617,22 +618,80 @@ class _ConstraintsDescriptorState extends State<ConstraintsDescriptor>
+    final Map<String, Object> flexDetails = node.diagnostic?.json['flex'];
+    final int numOfChildren = node.diagnostic.childrenNow.length;
+    final List<Widget> children = node.diagnostic.childrenNow.map((RemoteDiagnosticsNode child) => Expanded(
+      flex: 1,
```
you can now use List comprehensions for this sort of case in dart. so
[for (var child in node.diagnostic.childrenNow) 
   Container(something something)
];

jacob314 commented on packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart:
```diff
@@ -617,22 +618,80 @@ class _ConstraintsDescriptorState extends State<ConstraintsDescriptor>
+          children: [
+            Container(
+              margin: const EdgeInsets.only(bottom: 4.0),
+              child: const Text('Story of Your Flex Widget', style: TextStyle(
```
Add the name of the widget class. So:
'Story of the flex layout of your ${diagnostic.description} widget'
so
Story of the flex layout of your Column widget
or
Story of the flex layout of your Row widget.

jacob314 commented on packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart:
```diff
@@ -617,22 +618,80 @@ class _ConstraintsDescriptorState extends State<ConstraintsDescriptor>
+
   @override
   Widget build(BuildContext context) {
+    final Map<String, Object> flexDetails = node.diagnostic?.json['flex'];
```
Exciting to see this!
Go ahead and add getters for the extra json fields you add to RemoteDiagnosticsNode.

jacob314 commented on packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart:
```diff
@@ -617,22 +618,80 @@ class _ConstraintsDescriptorState extends State<ConstraintsDescriptor>
+            ),
+            Expanded(
+              child: Container(
+                color: Colors.blue,
```
try to use colors from the Theme whenever possible. That will make the ui work in dark and light theme.
See  calls to Theme.of in the code for examples.

Commit:
- Avoid using hardcoded color and instead use Color from Theme (to support Light/Dark theme)
- Remove hack in InspectorTree node
- Remove wrong import
- Rename 'Story of your Flex widget' to 'Story of the layout of your Row/Column widget'

albertusangga commented:
Check in 4:
- Avoid using hardcoded color and instead use Color from Theme (to support Light/Dark theme)
- Remove hack in InspectorTree node
- Remove wrong import
- Rename 'Story of your Flex widget' to 'Story of the layout of your Row/Column widget'

Screenshots:
![image](https://user-images.githubusercontent.com/25892469/67455092-4f655b00-f5e1-11e9-9b40-b4536d330e34.png)

Commit:
Added experimental flag for story of layout feature changes, add tabs for details tree, and several minor changes.

- Create proper deserializer method for flex properties from DiagnosticsNode
- Create new flag (`isExperimentalStoryOfLayoutEnabled`) in `InspectorTreeFlutterController` to prevent showing all this features changes until my PR (github.com/albertusangga/flutter/pull/2) is landed to Flutter main repository
- Inspector details tree now live inside of Tabs! User can focus on layout details instead if they want to.
- Removed animation from DianogsticsNodeDescriptions for now (due to minor performance concerns and adding new flag, might add it later!)
- Changed hardcoded constants for Buttons in InspectorScreen to prevent button groups expanding to two lines

albertusangga commented:
## Check in 5:

### Changelist:
- Create proper deserializer method for flex properties from DiagnosticsNode
- Create new flag (`isExperimentalStoryOfLayoutEnabled`) in `InspectorTreeFlutterController` to prevent showing all this features changes until my PR (github.com/albertusangga/flutter/pull/2) is landed to Flutter main repository
- Inspector details tree now live inside of Tabs! User can focus on layout details instead if they want to.
- Removed animation from DianogsticsNodeDescriptions for now (due to minor performance concerns and adding new flag, might add it later!)
- Changed hardcoded constants for Buttons in InspectorScreen to prevent button groups expanding to two lines 

### Screenshots:
- Inspector screen have tabs now!
![image](https://user-images.githubusercontent.com/25892469/67539758-02908b80-f698-11e9-91c2-1a160cda192b.png)

### TODOs:
- Remove `// constraints` word from ConstraintsDescriptor
- Highlight only unconstrained word in the ConstraintsDescriptor
- Move StoryOfYourFlexWidget from Dialog to Tab

Commit:
Remove debugLayoutModeEnabled from DiagnosticsNodeDescription

Commit:
Change isExperimentalStoryOfLayoutEnabled default value to false

Commit:
Revert diagnostics.dart

Commit:
Revert spreading getServiceExtensionWidgets() in Flutter InspectorScreen

Commit:
Remove unnecessary isSummaryTreeNode field in InspectorTreeNode constructor

Commit:
Display tab by default and fix code style and formatting

Commit:
Port story of flex widget to Tab

Commit:
Align expandCollapseButtons in the same row with the TabBar

Commit:
Merge branch 'master' of github.com:flutter/devtools into story-of-layout

Commit:
Rename layout_tab to inspector_screen_details_tab

Commit:
Partial highlight constraint and change deserialize method of FlexProperties

Commit:
Minor cleanups

- Stop exposing InspectorController in InspectorTreeControllerFlutter
- Remove shouldHighlightConstraint getter since it is unnecessary now
- Rename some variable names
- Add copyright in new files

Commit:
Add todos

Commit:
Add simple widget tests and handle case when Flex element does not have children

Commit:
Extended EnumDeserializer to contain reverse lookup and rename it to EnumUtils instead

Commit:
Cleanup & listen to InspectorController instead of TreeControllerFlutter

Commit:
Remove todo

Commit:
Remove InspectorControllerClient implementation in LayoutDetailsTab

Commit:
Rename deserializer to enum utils and make enum utils private

Commit:
Merge branch 'master' of github.com:flutter/devtools into story-of-layout

Commit:
Remove redundant method calls

Commit:
Stop passing InspectorController to InspectorTree flutter

- Add todos
- Rename toName & toEnum to name and enum
- Minor cleanups

Commit:
Add TODO

Commit:
Remove redundant comment

Commit:
Merge branch 'master' of github.com:flutter/devtools into story-of-layout

Commit:
Merge branch 'master' of github.com:flutter/devtools into story-of-layout

Commit:
Merge branch 'master' of github.com:flutter/devtools into story-of-layout

Commit:
Fix TabBar responsiveness

Commit:
Temp checkin (Arrows in progress)

Commit:
Add main/cross axis visualizer
